### PR TITLE
Feature/648 pathfinder docs

### PR DESCRIPTION
**Pull request by mitzimorris**
#### Submission Checklist

- [x] Builds locally 
- [ ] New functions marked with `` `r since("VERSION")` ``
- [x] Declare copyright holder and open-source license: see below

#### Summary

Added information on Pathfinder algorithm to Stan Reference Manual and CmdStan Guide.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC BY-ND 4.0 (https://creativecommons.org/licenses/by-nd/4.0/)

### CODE REVIEW

File: src/cmdstan-guide/pathfinder_intro.Rmd
Comment (WardBrian) [inline src/cmdstan-guide/pathfinder_intro.Rmd:27]: switch -- from what? I think we can just say "in order to use", if we think this paragraph is necessary. For the initial release it might even be confusing, since re-compiling old Stan programs is indeed necessary

File: src/cmdstan-guide/pathfinder_intro.Rmd
Comment (WardBrian) [inline src/cmdstan-guide/pathfinder_intro.Rmd:144]: I think we should add these citations to the `bibtex/all.bib` file and use them that way.

We also might want to consider using the official JMLR version, rather than the arxiv? It's still free to read, https://www.jmlr.org/papers/v23/21-0889.html

File: src/reference-manual/pathfinder.Rmd
Comment (WardBrian) [inline src/reference-manual/pathfinder.Rmd:24]: Same note on citation

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (WardBrian) [inline src/cmdstan-guide/pathfinder_config.Rmd:99]: The whitespace at the end of this file can be cleaned up

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (WardBrian) [inline src/cmdstan-guide/pathfinder_config.Rmd:1]: Because the diagnostic file structure is quite different for Pathfinder, I think it deserves its own heading on the detailed config page

**Review by WardBrian — CHANGES_REQUESTED** [review]
Mostly notes on wording or structure. I'll leave detailed comments on the description of the algorithm to those who understand it better, but to my level of knowledge it all sounds correct

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (mitzimorris) [reply src/cmdstan-guide/pathfinder_config.Rmd:1]: added

**Comment by mitzimorris** [comment]
ready for further review.

File: src/reference-manual/variational.Rmd
Comment (WardBrian) [inline src/reference-manual/variational.Rmd:1]: In the CmdStan Guide we updated the existing title of the pages on ADVI (rather than it just being called Variational Inference) - thoughts on doing the same in the reference manual, or moving Pathfinder to be a sub-page alongside ADVI in this section?

File: src/cmdstan-guide/pathfinder_intro.Rmd
Comment (WardBrian) [inline src/cmdstan-guide/pathfinder_intro.Rmd:25]: Sentence dies off - I'm guessing it was meant to end with something like "yields 1000 approximate draws" or something?

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (WardBrian) [inline src/cmdstan-guide/pathfinder_config.Rmd:58]: Should we just link to the page on the LBFGS config, rather than duplicate? I can see an argument either way

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (WardBrian) [inline src/cmdstan-guide/pathfinder_config.Rmd:85]: We should describe separately what happens if num_paths == 1. I think that we don't append anything to the requested csv name, but if we do we should just note that somewhere

**Review by WardBrian — COMMENTED** [review]
Thanks! A few more comments/suggestions from me

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (mitzimorris) [reply src/cmdstan-guide/pathfinder_config.Rmd:58]: yes, link because CmdStan interface does in fact reuse the argument objects.

File: src/cmdstan-guide/pathfinder_intro.Rmd
Comment (mitzimorris) [reply src/cmdstan-guide/pathfinder_intro.Rmd:25]: right - that was the point where I realized we needed to fix the default config in CmdStan.

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (mitzimorris) [reply src/cmdstan-guide/pathfinder_config.Rmd:85]: added.

File: src/bibtex/all.bib
Comment (bob-carpenter) [inline src/bibtex/all.bib:411]: The best reference here is JMLR from 2017:  https://www.jmlr.org/papers/volume18/16-107/16-107.pdf

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_config.Rmd:12]: I would add "and optionally importance resampled from several runs of Pathfinder".

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_config.Rmd:37]: These are all used by Pathfinder---the first lot get passed through to the embedded use of L-BFGS.

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_config.Rmd:69]: I don't think there's any reason to save the outputs of each single-path run other than perhaps for diagnostic purposes.

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_config.Rmd:70]: "between" -> "among"

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_config.Rmd:71]: "are" -> "is" (sets are singular)

or better, just "the resulting CSV files are:"

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_config.Rmd:85]: is this the order they're in?  parameters, etc. should come out in the order they are declared

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_config.Rmd:87]: I'd drop this or say each row represents a draw (and hence the whole file represents a sample).

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_config.Rmd:89]: These are not rows---they're comments.

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_config.Rmd:94]: remove redundant spaces

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_config.Rmd:97]: Presumably if this doesn't show up no output is produced?

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_config.Rmd:80]: line 80 to "These CSV files have the following structure:"

File: src/cmdstan-guide/pathfinder_intro.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_intro.Rmd:3]: This feels redundant with [src/cmdstan-guide/pathfinder_config.Rmd](https://github.com/stan-dev/docs/pull/651/files#diff-f498b1eea44ea4ceffd1f8e28e7b1953b2df2f2566f6de47bd499ae561438f14), so I would just remove the earlier description and put a pointer to here instead.

File: src/cmdstan-guide/pathfinder_intro.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_intro.Rmd:14]: "stochastic gradient descent (SGD)" should continue with: 

on the Monte Carlo computation of the evidence lower bound (ELBO).

File: src/cmdstan-guide/pathfinder_intro.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_intro.Rmd:15]: This is too unqualified.  The instability comes from the stochastic objective, not from gradient descent vs. L-BFGS.  I'd say "Pathfinder's approach is both faster and more stable than that of ADVI."

File: src/cmdstan-guide/pathfinder_intro.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_intro.Rmd:21]: running multiple Pathfinder runs -> running multiple Pathfinder instances

File: src/cmdstan-guide/pathfinder_intro.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_intro.Rmd:25]: I think you should continue after "draws" with ", which are then importance resampled down to 100 final draws."

File: src/cmdstan-guide/pathfinder_intro.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_intro.Rmd:27]: This also feels redundant with the config section.

File: src/cmdstan-guide/pathfinder_intro.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_intro.Rmd:89]: Whare are "StanCSV files"?  Can't we just say "CSV" files. 

"one file which" -> "one file that" [which is for non-restrictive relatives!]

"one file which contains" -> "one file per single-path Pathfinder run" (as written it feels like one file)

File: src/cmdstan-guide/pathfinder_intro.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_intro.Rmd:92]: + the default number of single-path runs

File: src/cmdstan-guide/pathfinder_intro.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_intro.Rmd:105]: pathfinder -> Pathfinder

File: src/cmdstan-guide/pathfinder_intro.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_intro.Rmd:111]: remove line 111

File: src/cmdstan-guide/pathfinder_intro.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_intro.Rmd:117]: I'd remove 117, 121, 126, and 129, even though they're in the output file.  It's just comments.

File: src/reference-manual/pathfinder.Rmd
Comment (bob-carpenter) [inline src/reference-manual/pathfinder.Rmd:3]: Drop line 3---we've avoided talking about interface specifics in the reference manual.  Just say that "Stan supports the Pathfinder algorithm of @zhang..."

File: src/reference-manual/pathfinder.Rmd
Comment (bob-carpenter) [inline src/reference-manual/pathfinder.Rmd:19]: Unless we can parallelize, I'd drop 19--22.

**Review by bob-carpenter — CHANGES_REQUESTED** [review]
A couple dozen nitpicky comments, but otherwise this looks fine.  At some point we need to be a bit more precise in the reference manual algorithm descriptions.  And I'd like to move all the algorithm config description there.  But that's a much bigger PR than this one.  I think this one's OK as is.

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (mitzimorris) [reply src/cmdstan-guide/pathfinder_config.Rmd:69]: this is documenting current behavior.  we could change that.

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (mitzimorris) [reply src/cmdstan-guide/pathfinder_config.Rmd:70]: reworded.

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (mitzimorris) [reply src/cmdstan-guide/pathfinder_config.Rmd:97]: added clarification

File: src/cmdstan-guide/pathfinder_intro.Rmd
Comment (mitzimorris) [reply src/cmdstan-guide/pathfinder_intro.Rmd:25]: you mean 1000, right?  changed.

File: src/cmdstan-guide/pathfinder_intro.Rmd
Comment (mitzimorris) [reply src/cmdstan-guide/pathfinder_intro.Rmd:27]: not everyone will read both sections.

File: src/cmdstan-guide/pathfinder_intro.Rmd
Comment (mitzimorris) [reply src/cmdstan-guide/pathfinder_intro.Rmd:89]: the CSV format is underspecified.  these files contain comments before the header row and after the data rows; not all CSV parsers will be happy. 
linking to Pathfinder CSV section

**Comment by mitzimorris** [comment]
made changes, ready for re-review or approval.

**Review by WardBrian — APPROVED** [review]
I believe all my comments were adequately addressed so I'm approving to un-block my "Request Changes". I think @bob-carpenter also needs to approve before the merge button unlocks again

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (bob-carpenter) [reply src/cmdstan-guide/pathfinder_config.Rmd:60]: I didn't realize.  Would it be possible to default to not outputting those files?  The issue is that it's like a factor of N - 1 times as much output where N is the number of chains.

Comment (mitzimorris) [reply src/cmdstan-guide/pathfinder_config.Rmd:60]: yes we could do that.  will open an issue on CmdStan.

**Comment by mitzimorris** [comment]
@bob-carpenter and @WardBrian please re-review.  updated docs for changes introduced by CmdStan PR https://github.com/stan-dev/cmdstan/pull/1184

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (mitzimorris) [reply src/cmdstan-guide/pathfinder_config.Rmd:167]: you're right.  will reword

Comment (mitzimorris) [reply src/cmdstan-guide/pathfinder_config.Rmd]: done!

**Review by WardBrian — APPROVED** [review]

File: src/bibtex/all.bib
Comment (bob-carpenter) [inline src/bibtex/all.bib:1779]: Missing final newline.

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_config.Rmd:29]: This should move down to go with the section on L-BFGS config.

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_config.Rmd:34]: I think it would help to mention this is the final number of draws returned by multi-path pathfinder.

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_config.Rmd:41]: I'd drop the parens or replace with braces (\{ and \}---they need to be escaped) as it looks too much like a tuple this way.

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_config.Rmd:46]: This should then note that it's not the final number of draws for multi-path.

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_config.Rmd:56]: You should also mention it's the same arguments and the same defaults as the args aren't listed anywhere I see.  It might help to call these out even if you don't repeat the doc.

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_config.Rmd:168]: If we run just a single run of single-path pathfinder will it still importance resample?   I think it's OK either way, but should be documented.  Coming at this from the paper, it's not entirely clear how to configure single-path pathfinder, not that I'd recommend anyone doing that.

File: src/cmdstan-guide/pathfinder_intro.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_intro.Rmd:8]: It's negative inverse Hessian.  I think I wrote that wrong elsewhere.

File: src/cmdstan-guide/pathfinder_intro.Rmd
Comment (bob-carpenter) [inline src/cmdstan-guide/pathfinder_intro.Rmd:93]: There are a lot of stray newlines here I'd just collapse down to max 80 chars/line.

**Review by bob-carpenter — CHANGES_REQUESTED** [review]
looks good---just a few minor comments

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (mitzimorris) [reply src/cmdstan-guide/pathfinder_config.Rmd:34]: added more text to the beginning of this section to clarify behavoir.

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (mitzimorris) [reply src/cmdstan-guide/pathfinder_config.Rmd:56]: reworded to clarify.

File: src/cmdstan-guide/pathfinder_config.Rmd
Comment (mitzimorris) [reply src/cmdstan-guide/pathfinder_config.Rmd:168]: no, single path pathfinder doesn't importance resample.  there are two service methods; for `num_paths=1`,  service `pathfinder_lbfgs_single` is run.

Comment (mitzimorris) [reply src/cmdstan-guide/pathfinder_config.Rmd:168]: reworded to clarify

File: src/cmdstan-guide/pathfinder_intro.Rmd
Comment (mitzimorris) [reply src/cmdstan-guide/pathfinder_intro.Rmd:93]: reformatted this file so that all prose is < 80 chars/line.

File: src/cmdstan-guide/pathfinder_intro.Rmd
Comment (mitzimorris) [reply src/cmdstan-guide/pathfinder_intro.Rmd:8]: changed here and in ref manual.

**Review by bob-carpenter — APPROVED** [review]
thanks---looks good

**Pull request merged**
